### PR TITLE
Models without Vocabulary

### PR DESCRIPTION
**Pull request by Xarbirus**
I made some changes to the model converter so that it could create a gguf model without a built-in dictionary.
This will allow to use any custom external dictionary in an application built with llama.cpp.

### CODE REVIEW

File: llama.cpp
Comment (cebtenzzre) [inline llama.cpp:1]: This commit seems like it belongs in a separate PR.

File: llama.h [hunk 1]
```diff
@@ -632,6 +635,8 @@ extern "C" {
     // n_threads is the number of threads used for generation (single token)
     // n_threads_batch is the number of threads used for prompt and batch processing (multiple tokens)
     LLAMA_API void llama_set_n_threads(struct llama_context * ctx, uint32_t n_threads, uint32_t n_threads_batch);
+    // Get the number of threads used for decoding
+    LLAMA_API void llama_get_n_threads(struct llama_context * ctx, uint32_t * n_threads, uint32_t * n_threads_batch);
```
Comment (cebtenzzre) [inline llama.h:639]: Some of these changes seem like they belong in a separate PR.

Comment (Xarbirus) [reply llama.h]: Removed.

File: convert.py [hunk 2]
```diff
@@ -1322,26 +1357,33 @@ def _create_special_vocab(self, vocab: Vocab, vocabtype: str, model_parent_path:
...
+            return SentencePieceVocab(
                 path, added_tokens_path if added_tokens_path.exists() else None
             )
         elif vocabtype == "hfft":
-            vocab = HfVocab(
+            return HfVocab(
                 path, added_tokens_path if added_tokens_path.exists() else None
             )
         else:
             raise ValueError(f"Unsupported vocabulary type {vocabtype}")
```
Comment (cebtenzzre) [inline convert.py:1378]: Change the `elif`s here to `if`s, and remove the `else` from the `raise`.

**Review by ggerganov — APPROVED** [review]
This seems something that can be useful

**Comment by Xarbirus** [comment]
@cebtenzzre right now we're using some kind of [this](https://github.com/VKCOM/YouTokenToMe) tokenizer with the llama model trained by our ml engineers. And in our system the vocab is on the client side, and the server only processes tokens. So there is no need for the vocab to be included in the model.

File: convert.py (hunk 2 above)
Comment (Xarbirus) [reply convert.py]: fixed

File: convert.py [hunk 3]
```diff
@@ -385,6 +385,12 @@ def all_tokens(self) -> Iterable[tuple[bytes, float, gguf.TokenType]]:
         yield from self.bpe_tokens()
         yield from self.added_tokens()
 
+    def get_tokenizer_model(self) -> str:
+        return "gpt2"
+
+    def get_name(self) -> str:
+        return "bpe"
```
Comment (cebtenzzre) [inline convert.py:392]: Getter functions are not pythonic. These should probably just be class attributes, e.g. the top of the class would look like:
```python
class BpeVocab:
    tokenizer_model = "gpt2"
    name = "bpe"

    def __init__(...):
```

File: convert.py [hunk 4]
```diff
@@ -1131,24 +1169,33 @@ def write_all(
...
+        for name, lazy_tensor in model.items():
+            of.add_tensor_info(name, lazy_tensor)
+
+        of.write_meta()
+        of.write_tensor_info()
+
+        # tensor data
+        of.write_tensor_data(ftype, model, concurrency)
 
         of.close()
```
Comment (cebtenzzre) [inline convert.py:1200]: I know that there are already a lot of parameters to write_all, but there would be less code duplication if you just made `no_vocab` a boolean parameter instead of making a separate function for it.

File: llama.h [hunk 5]
```diff
@@ -59,9 +59,10 @@ extern "C" {
     typedef int32_t llama_seq_id;
 
     enum llama_vocab_type {
-        LLAMA_VOCAB_TYPE_SPM = 0, // SentencePiece
-        LLAMA_VOCAB_TYPE_BPE = 1, // Byte Pair Encoding
-        LLAMA_VOCAB_TYPE_WPM = 2, // WordPiece
+        LLAMA_VOCAB_TYPE_SPM      = 0, // SentencePiece
+        LLAMA_VOCAB_TYPE_BPE      = 1, // Byte Pair Encoding
+        LLAMA_VOCAB_TYPE_WPM      = 2, // WordPiece
+        LLAMA_VOCAB_TYPE_NO_VOCAB = 3, // For models without vocab
```
Comment (cebtenzzre) [inline llama.h:65]: Maybe we should call this LLAMA_VOCAB_TYPE_NONE instead.

File: convert.py (hunk 3 above)
Comment (Xarbirus) [reply convert.py]: Done.

File: convert.py (hunk 4 above)
Comment (Xarbirus) [reply convert.py]: I decided not to add one more argument to the `write_all` function, but to check the name of the `vocab`. WDYT?

File: llama.h (hunk 5 above)
Comment (Xarbirus) [reply llama.h]: Used `LLAMA_VOCAB_TYPE_NONE = 0`. I hope I won't break compatibility.

File: convert.py [hunk 6]
```diff
@@ -931,13 +948,19 @@ def bounded_parallel_map(func: Callable[[In], Out], iterable: Iterable[In], conc
...
         raise ValueError(
-            f"The model's vocab size is set to -1 in params.json. Please update it manually. Maybe {vocab.vocab_size}?"
+            f"The model's vocab size is set to -1 in params.json. Please update it manually.{f' Maybe {vocab.vocab_size}?' if hasattr(vocab, 'vocab_size') else ''}"
         )
 
+
+def prepare_vocab(params: Params, vocab: Vocab, pad_vocab: bool = False) -> None:
+    check_vocab_size(params, vocab)
+    if vocab.name == "no_vocab":
+        return
```
Comment (cebtenzzre) [inline convert.py:962]:
```suggestion
    if isinstance(vocab, NoVocab):
        return  # model has no vocab
```

File: convert.py [hunk 7]
```diff
@@ -1114,14 +1141,17 @@ def write_all(
...
-        check_vocab_size(params, vocab, pad_vocab=pad_vocab)
+        prepare_vocab(params, vocab, pad_vocab=pad_vocab)
 
         of = OutputFile(fname_out, endianess=endianess)
 
         # meta data
         of.add_meta_arch(params)
-        of.add_meta_vocab(vocab)
-        of.add_meta_special_vocab(svocab)
+        if vocab.name == "no_vocab":
```
Comment (cebtenzzre) [inline convert.py:1150]:
```suggestion
        if isinstance(vocab, NoVocab):
```

File: convert.py (hunk 6 above)
Comment (cebtenzzre) [reply convert.py:962]: And since check_vocab_size is no longer called anywhere, isn't this effectively an unnecessary rename? You should make prepare_vocab part of check_vocab_size again.

Comment (Xarbirus) [reply convert.py]: Done.

File: convert.py [hunk 8]
```diff
@@ -1013,20 +1033,6 @@ def add_meta_arch(self, params: Params) -> None:
...
-        }.get(type(vocab))
-
-        # Block if vocab type is not predefined
-        if tokenizer_model is None:
-            raise ValueError("Unknown vocab type: Not supported")
-
-        return tokenizer_model
-
     def extract_vocabulary_from_model(self, vocab: Vocab) -> tuple[list[bytes], list[float], list[gguf.TokenType]]:
         tokens = []
```
Comment (cebtenzzre) [inline convert.py:1037]: mypy would like you to add an assertion in this function to make certain that vocab is not a NoVocab. Otherwise, the all_tokens and vocab_size attributes will not be available.
```suggestion
        assert not isinstance(vocab, NoVocab)
        tokens = []
```

Comment (Xarbirus) [reply convert.py:1039]: Done

**Review by cebtenzzre — APPROVED** [review]
